File: go/pondsocket/store.go
package pondsocket

import (
	"sync"
)

type store[T any] struct {
	mutex sync.RWMutex
	store map[string]T
}

func newStore[T any]() *store[T] {
	return &store[T]{
		mutex: sync.RWMutex{},
		store: make(map[string]T),
	}
}

func (s *store[T]) Create(key string, value T) error {
	s.mutex.Lock()

	defer s.mutex.Unlock()

	if _, exists := s.store[key]; exists {
		return conflict(key, "Key already exists")
	}
	s.store[key] = value
	return nil
}

func (s *store[T]) Read(key string) (T, error) {
	s.mutex.RLock()

	defer s.mutex.RUnlock()

	var zeroValue T
	value, exists := s.store[key]
	if !exists {
		return zeroValue, notFound(key, "Key does not exist")
	}
	return value, nil
}

func (s *store[T]) Update(key string, value T) error {
	s.mutex.Lock()

	defer s.mutex.Unlock()

	if _, exists := s.store[key]; !exists {
		return notFound(key, "Key does not exist")
	}
	s.store[key] = value
	return nil
}

func (s *store[T]) Delete(key string) error {
	s.mutex.Lock()

	defer s.mutex.Unlock()

	if _, exists := s.store[key]; !exists {
		return notFound(key, "Key does not exist")
	}
	delete(s.store, key)

	return nil
}

func (s *store[T]) List() map[string]T {
	s.mutex.RLock()

	defer s.mutex.RUnlock()

	result := make(map[string]T)

	for key, value := range s.store {
		result[key] = value
	}
	return result
}

func (s *store[T]) Keys() *array[string] {
	s.mutex.RLock()

	defer s.mutex.RUnlock()

	keys := newArray[string]()

	for key := range s.store {
		keys.push(key)
	}
	return keys
}

func (s *store[T]) Values() *array[T] {
	s.mutex.RLock()

	defer s.mutex.RUnlock()

	values := newArray[T]()

	for _, value := range s.store {
		values.push(value)
	}
	return values
}

func (s *store[T]) GetByKeys(keys ...string) *array[T] {
	s.mutex.RLock()

	defer s.mutex.RUnlock()

	values := newArray[T]()

	for _, key := range keys {
		if value, exists := s.store[key]; exists {
			values.push(value)
		}
	}
	return values
}

func (s *store[T]) Len() int {
	s.mutex.RLock()

	defer s.mutex.RUnlock()

	return len(s.store)
}

File: python/pondsocket/src/pondsocket/store.py
from __future__ import annotations

import asyncio
from typing import Generic, TypeVar

from .errors import conflict, not_found

T = TypeVar("T")


class Store(Generic[T]):
    __slots__ = ("_data", "_lock")

    def __init__(self) -> None:
        self._lock = asyncio.Lock()
        self._data: dict[str, T] = {}

    async def create(self, key: str, value: T) -> None:
        async with self._lock:
            if key in self._data:
                raise conflict(key, "Key already exists")
            self._data[key] = value

    async def read(self, key: str) -> T:
        async with self._lock:
            if key not in self._data:
                raise not_found(key, "Key does not exist")
            return self._data[key]

    async def get(self, key: str, default: T | None = None) -> T | None:
        async with self._lock:
            return self._data.get(key, default)

    async def has(self, key: str) -> bool:
        async with self._lock:
            return key in self._data

    async def update(self, key: str, value: T) -> None:
        async with self._lock:
            if key not in self._data:
                raise not_found(key, "Key does not exist")
            self._data[key] = value

    async def upsert(self, key: str, value: T) -> None:
        async with self._lock:
            self._data[key] = value

    async def delete(self, key: str) -> None:
        async with self._lock:
            if key not in self._data:
                raise not_found(key, "Key does not exist")
            del self._data[key]

    async def discard(self, key: str) -> bool:
        async with self._lock:
            if key not in self._data:
                return False
            del self._data[key]
            return True

    async def snapshot(self) -> dict[str, T]:
        async with self._lock:
            return dict(self._data)

    async def keys(self) -> list[str]:
        async with self._lock:
            return list(self._data.keys())

    async def values(self) -> list[T]:
        async with self._lock:
            return list(self._data.values())

    async def get_many(self, *keys: str) -> list[T]:
        async with self._lock:
            return [self._data[k] for k in keys if k in self._data]

    async def len(self) -> int:
        async with self._lock:
            return len(self._data)

    async def clear(self) -> None:
        async with self._lock:
            self._data.clear()
